subprojects/core/src/main/java/org/gradle/internal/model/CalculatedValueContainerCache.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.model;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Maps;
+import org.gradle.api.NonNullApi;
+import org.gradle.internal.DisplayName;
+
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * This class manages the coordination between two or more threads that are trying to calculate the same value
+ * concurrently. It ensures that only one value container will be used for concurrent calculations for the same key,
+ * and that the value is only calculated once for concurrent calls.  Note that this only applies to concurrent
+ * calculations of the same value.  If two threads are trying to calculate the values serially, then each will
+ * get their own value container.  In other words, this class only ensures that concurrent threads always get
+ * the same value container, not that the same value container will be used every time
+ * {@link #getReference(DisplayName, Supplier)} is called for a given key.
+ *
+ * Also note that this class is not appropriate for concurrent threads providing different suppliers for the same key.
+ * In that case, the first supplier to be called will be used when calculating the value and then other threads will
+ * get the same value container with the cached value from the first supplier.  In other words, their suppliers will
+ * never be called because the value will already be cached.
+ *
+ * @param <T> the type to use for the uniquely identifiable key
+ * @param <V> the type of the value that the value containers will hold
+ */
+@NonNullApi
+public class CalculatedValueContainerCache<T extends DisplayName, V> {
```

Comment:
Instead of this, could we just use a Guava cache with an eviction policy or weak values?

Reply:
Yeah, that was my feeling, too.  The only concern might be the heap overhead, but I would expect weak values to keep that in check.  I've changed this and kicked off some perf tests to see if it has any negative effect.